core/src/main/scala/org/http4s/Headers.scala

```diff
@@ -1,158 +1,199 @@
-        case hs: Headers => hs.asInstanceOf[That]
-        case hs => new Headers(hs.toList.asInstanceOf[List[Header]]).asInstanceOf[That]
-      } else {
-        val hs = that.toList.asInstanceOf[List[Header]]
-        val acc = new ListBuffer[Header]
-        this.headers.foreach { orig =>
-          orig.parsed match {
-            case _: Header.Recurring => acc += orig
-            case _: `Set-Cookie` => acc += orig
-            case h if !hs.exists(_.name == h.name) => acc += orig
-            case _ => // NOOP, drop non recurring header that already exists
-          }
-        }
-
-        val h = new Headers(acc.prependToList(hs))
-        h.asInstanceOf[That]
-      }
-    } else super.++(that)
-
-  override def hashCode(): Int = this.headers.hashCode()
-
-  override def equals(that: Any): Boolean = that match {
-    case otherheaders: Headers =>
-      this.toList.equals(otherheaders.toList)
-    case _ => false
+// import scala.collection.{GenTraversableOnce, immutable, mutable}
+// import scala.collection.generic.CanBuildFrom
+// import scala.collection.mutable.ListBuffer
+
+
+final case class Headers(raw: List[Header.RawHeader]){
+  def get[A](implicit A: Header[A]): Option[A] = {
+    if (A.isSingleton) {
+      raw.find(_.name == A.name).flatMap(h => A.parse(h.value))
+    } else {
+      val values = raw.filter(_.name == A.name).map(_.value)
+      if (values.nonEmpty) A.parse(values.mkString(", "))
+      else None
+    }
   }
 
-  /** Removes the `Content-Length`, `Content-Range`, `Trailer`, and
-    * `Transfer-Encoding` headers.
-    *
-    *  https://tools.ietf.org/html/rfc7231#section-3.3
-    */
-  def removePayloadHeaders: Headers =
-    filterNot(h => Headers.PayloadHeaderKeys(h.name))
-
-  def redactSensitive(
-      redactWhen: CaseInsensitiveString => Boolean = Headers.SensitiveHeaders.contains): Headers =
-    headers.map {
-      case h if redactWhen(h.name) => Header.Raw(h.name, "<REDACTED>")
-      case h => h
+  def getRepeatable(key: CaseInsensitiveString): Option[Header.RawHeader] = {
+      val hs = raw.filter(_.name == key)
+      if (hs.isEmpty) None
+      else {
+        Some(Header.RawHeader(key, hs.mkString(", ")))
+      }
     }
-}
-
-object Headers {
-  val empty = apply()
 
-  /** Create a new Headers collection from the headers */
-  def apply(headers: Header*): Headers = Headers(headers.toList)
+    def getSingleton(key: CaseInsensitiveString): Option[Header.RawHeader] =
+      raw.find(_.name == key)
 
-  /** Create a new Headers collection from the headers */
-  def apply(headers: List[Header]): Headers = new Headers(headers)
-
-  implicit val canBuildFrom: CanBuildFrom[Traversable[Header], Header, Headers] =
-    new CanBuildFrom[TraversableOnce[Header], Header, Headers] {
-      def apply(from: TraversableOnce[Header]): mutable.Builder[Header, Headers] = newBuilder
-
-      def apply(): mutable.Builder[Header, Headers] = newBuilder
+    def put(headers0: Sigma[Header]*): Headers = {
```

Comment:
This is an interesting technique that would make an interesting Typelevel blog post if anyone is so motivated.